In [32]:
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from scipy.stats import pearsonr
import matplotlib.pyplot as plt

In [33]:
# Importação dos conjutos de amostras
dados_treino = pd.read_csv("conjunto_de_treinamento.csv", delimiter=",", decimal=".")
dados_teste = pd.read_csv("conjunto_de_teste.csv", delimiter=",", decimal=".")

In [55]:
dados_teste.head()

id_solicitante  produto_solicitado  dia_vencimento forma_envio_solicitacao  \
0           20001                   1              25              presencial   
1           20002                   1              10                internet   
2           20003                   1              10                internet   
3           20004                   1              10              presencial   
4           20005                   1              10              presencial   

   tipo_endereco sexo  idade  estado_civil  qtde_dependentes  grau_instrucao  \
0              1    M     37             2                 0               0   
1              1    F     31             2                 0               0   
2              1    F     18             2                 0               0   
3              1    F     55             2                 0               0   
4              1    F     55             1                 0               0   

   ...  estado_onde_trabalha possui_telefone_trabalho  \
0  ...                                              N   
1  ...                    RJ                        N   
2  ...                    RS                        N   
3  ...                                              N   
4  ...                                              N   

  codigo_area_telefone_trabalho meses_no_trabalho profissao  ocupacao  \
0                                               0       0.0       0.0   
1                                               0       9.0       5.0   
2                                               0       9.0       2.0   
3                                               0       9.0       1.0   
4                                               0       0.0       1.0   

   profissao_companheiro grau_instrucao_companheiro  local_onde_reside  \
0                    0.0                        0.0              384.0   
1                    NaN                        NaN              275.0   
2                    NaN                        NaN              948.0   
3                    0.0                        0.0              581.0   
4                    0.0                        0.0              573.0   

   local_onde_trabalha  
0                384.0  
1                275.0  
2                948.0  
3                581.0  
4                573.0  

[5 rows x 41 columns]

In [57]:
dados_treino.head()

id_solicitante  produto_solicitado  dia_vencimento forma_envio_solicitacao  \
0               1                   1              10              presencial   
1               2                   1              25                internet   
2               3                   1              20                internet   
3               4                   1              20                internet   
4               5                   7               1                internet   

   tipo_endereco sexo  idade  estado_civil  qtde_dependentes  grau_instrucao  \
0              1    M     85             2                 0               0   
1              1    F     38             1                 0               0   
2              1    F     37             2                 0               0   
3              1    M     37             1                 1               0   
4              1    F     51             1                 3               0   

   ...  possui_telefone_trabalho codigo_area_telefone_trabalho  \
0  ...                         N                                 
1  ...                         N                                 
2  ...                         N                                 
3  ...                         Y                            54   
4  ...                         N                                 

  meses_no_trabalho profissao ocupacao  profissao_companheiro  \
0                 0       9.0      1.0                    0.0   
1                 0       2.0      5.0                    NaN   
2                 0       NaN      NaN                    NaN   
3                 0       9.0      2.0                    NaN   
4                 0       9.0      5.0                    NaN   

   grau_instrucao_companheiro local_onde_reside  local_onde_trabalha  \
0                         0.0             600.0                600.0   
1                         NaN             492.0                492.0   
2                         NaN             450.0                450.0   
3                         NaN             932.0                932.0   
4                         NaN             440.0                440.0   

   inadimplente  
0             0  
1             0  
2             1  
3             1  
4             1  

[5 rows x 42 columns]

Antes de iniciarmos as operações é necessário atribuir valores numétricos às colunas com letras

In [58]:
colunas = dados_treino.columns

for coluna in colunas:
    dados_treino[coluna] = pd.to_numeric(dados_treino[coluna])
    dados_teste[coluna] = pd.to_numeric(dados_teste[coluna])

ValueError: Unable to parse string "presencial" at position 0

In [53]:
# Verifica perda de dados de treino ao utilizar somente linhas sem dados faltantes
# Testar substituindo os dados faltantes do treino pela média

linha_treino_com_na = dados_treino.shape[0]
linha_treino_sem_na = dados_treino.dropna().shape[0]

print("Perda de %0.2f%% dos dados de treino ao retirar linhas com informações faltantes." %(100 - linha_treino_sem_na*100/linha_treino_com_na))

Perda de 65.12% dos dados de treino ao retirar linhas com informações faltantes


A perda de 65.12% do dataset de treino pode representar um grande problema pois retirar grande parte da visão de todo o universo das amostrar. Para prosseguir sem perder essas amostras, substitui-se aos NaN pela média da coluna.

In [ ]:
quantificacao_atributos_nao_numericos = {
    'Y':1,
    'N':0,
    'M':0,
    'F':1,
    'N':2,
    'internet':0,
    'correios':1,
    'presencial':2,
    
}

for coluna in colunas:
    if 

In [27]:
#separacao em alvo e features
dados_treino = dados_treino.dropna()
x_treino = dados_treino.iloc[:, 1:-1]
y_treino = dados_treino.iloc[:, -1]

dados_teste = dados_teste.dropna()
x_teste = dados_teste.iloc[:, 1:-1]
y_teste = dados_teste.iloc[:, -1]

In [29]:
for numero_kneighbors in range(1, 40):
    classificador = KNeighborsClassifier(n_neighbors=numero_kneighbors, weights='uniform')
    
    classificador = classificador.fit(x_treino, y_treino)
    y_resposta = classificador.predict(x_teste)
    
    total   = len(y_teste)
    acertos = sum(y_resposta==y_teste)
    erros   = sum(y_resposta!=y_teste)

    print ("Total de amostras: " , total)
    print ("Respostas corretas:" , acertos)
    print ("Respostas erradas: " , erros)

    acuracia = acertos / total

    print ("Acurácia = %.1f %%" % (100*acuracia))

ValueError: could not convert string to float: 'presencial'